parsl/dataflow/dflow.py

```diff
@@ -379,38 +397,83 @@ def handle_exec_update(self, task_record, future):
             self.launch_if_ready(task_record)
 
     def handle_join_update(self, task_record, inner_app_future):
-        # Use the result of the inner_app_future as the final result of
-        # the outer app future.
-
-        # There is no retry handling here: inner apps are responsible for
-        # their own retrying, and joining state is responsible for passing
-        # on whatever the result of that retrying was (if any).
-
-        outer_task_id = task_record['id']
+        with task_record['join_lock']:
+            # inner_app_future has completed, which is one (potentially of many)
+            # futures the outer task is joining on.
+
+            # If the outer task is joining on a single future, then
+            # use the result of the inner_app_future as the final result of
+            # the outer app future.
+
+            # If the outer task is joining on a collection of futures, then
+            # check if the collection is all done, and if so, return a list
+            # of the results. Otherwise, this callback can do nothing and
+            # processing will happen in another callback (on the final Future
+            # to complete)
+
+            # There is no retry handling here: inner apps are responsible for
+            # their own retrying, and joining state is responsible for passing
+            # on whatever the result of that retrying was (if any).
+
+            outer_task_id = task_record['id']
+            logger.debug(f"Join callback for task {outer_task_id}, inner_app_future {inner_app_future}")
+
+            if task_record['status'] != States.joining:
+                logger.debug(f"Join callback for task {outer_task_id} skipping because task is not in joining state")
+                return
+
+            joinable = task_record['joins']  # Future or collection of futures
+
+            if isinstance(joinable, list):  # TODO more generic type than list?
+                for future in joinable:
+                    if not future.done():
+                        logger.debug(f"A joinable future {future} is not done for task {outer_task_id} - skipping callback")
+                        return  # abandon this callback processing if joinables are not all done
+
+            # now we know each joinable Future is done
+            # so now look for any exceptions
+            e = None
+            if isinstance(joinable, Future):
+                if joinable.exception():
+                    e = joinable.exception()
+            elif isinstance(joinable, list):
+                for future in joinable:
+                    if future.exception():
+                        e = future.exception()
```

Comment:
Not sure about this. The futures in the list are ordered, so I wonder if we'll lose info on all exceptions after the first one in the list?

Reply:
I added a new error, `JoinError`, which looks similar to DependencyError - it captures all of the exceptions and tids for failed joining apps.